app/src/main/java/com/sschuraytz/charcoaldrawing/UndoRedo.java

```diff
@@ -8,10 +8,13 @@
 public class UndoRedo {
 
     private Stack<Bitmap> currentStack = new Stack<>();
+    private Stack<Bitmap> undoneStack = new Stack<>();
     private Canvas bitmapCanvas;
+    private UndoRedoListener undoRedoListener;
```

Comment:
This is always null because it's never instantiated. The way to instantiate interfaces is by using the class that implements the interface. But, I'm pretty sure I don't want to do that with the MainActivity, which is where I currently implement the interface. 

Did you have a different activity in mind when you wrote "Then you'll implement an UndoRedoListener in your Activity."?

Reply:
I don't understand your question. It isn't **always** null, that's why you have a setter method. 

It looks like you have all this correct, but either I missed it or you aren't using the setter.